quit/web/modules/endpoint.py

```diff
@@ -119,44 +146,75 @@ def provenance():
-                res = graph.query(q)
-
-                if mimetype in ['text/html', 'application/xhtml_xml', '*/*']:
-                    results = res.serialize(format='html')
-                    response = make_response(render_template(
-                        "results.html", results=Markup(results.decode())))
-                    response.headers['Content-Type'] = 'text/html'
-                    return response
-                elif mimetype in ['application/json', 'application/sparql-results+json']:
-                    response = make_response(res.serialize(format='json'), 200)
-                    response.headers['Content-Type'] = 'application/json'
-                    return response
-                elif mimetype in [
-                    'application/rdf+xml', 'application/xml', 'application/sparql-results+xml'
-                ]:
-                    response = make_response(res.serialize(format='xml'), 200)
-                    response.headers['Content-Type'] = 'application/rdf+xml'
-                    return response
-                elif mimetype in ['application/csv', 'text/csv']:
-                    response = make_response(res.serialize(format='csv'), 200)
-                    response.headers['Content-Type'] = 'text/csv'
-                    return response
+            if mimetype == 'text/html':
+                return "Unsupported Query Type: <pre>{}</pre>".format(traceback.format_exc()), 400
             else:
-                raise UnSupportedQueryType()
-
+                return make_response('', 400)
         except Exception as e:
             logger.exception(e)
-            return "<pre>{}</pre>".format(traceback.format_exc()), 400
+
+            if mimetype == 'text/html':
+                return "<pre>{}</pre>".format(traceback.format_exc()), 400
+            else:
+                return make_response('', 400)
+
+        if query_type in ['SELECT', 'ASK']:
+            querytype = 'query'
+            res = graph.query(q)
+
+            # Set default
+            if mimetype == '*/*':
+                mimetype = 'application/sparql-results+xml'
+
+        elif query_type in ['CONSTRUCT', 'DESCRIBE']:
+            querytype = 'construct'
+            res = graph.query(q)
+
+            # Set default
+            if mimetype == '*/*':
+                mimetype = 'text/turtle'
+
+        else:
+            res = graph.update(q)
+
+            ref = request.values.get('ref', None) or default_branch
+            ref = 'refs/heads/{}'.format(ref)
+            quit.commit(
+                graph, res.get('delta', None), 'New Commit from QuitStore',
+                branch_or_ref, ref, query=q
+            )
+            return '', 200
+
+        return create_result_response(res, querytype, mimetype)
     else:
-        return render_template('provenance.html')
+        if mimetype == 'text/html':
+            return render_template('provenance.html')
+
+
+def create_result_response(res, querytype, mimetype):
```

Comment:
change signature to `create_result_response(res, mimetype, mimetypes[mimetype])`